core/src/main/scala/caliban/rendering/Renderer.scala

```diff
@@ -0,0 +1,559 @@
+    def nlOrNot = if (newline) writer += '\n' else writer
+
+    description.foreach {
+      case value if value.contains('\n') =>
+        val valueEscaped = value.replace(tripleQuote, s"\\$tripleQuote")
+        // check if it ends in quote but it is already escaped
+        if (value.endsWith("\\\"")) {
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrNot
+          writer ++= tripleQuote
+        } else if (value.last == '"') {
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrSp
+          writer ++= tripleQuote
+          // check if it ends in quote. We need to break the sequence of 4 or more '"'
+        } else {
+          // ok. No quotes at the end of value
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrNot
+          writer ++= tripleQuote
+        }
+      case value                         =>
+        if (newline) pad(indent, writer)
+        unsafeFastString(value, writer)
+        nlOrSp
+    }
+  }
+
+  private[caliban] lazy val fieldDefinitionRenderer: Renderer[FieldDefinition] = new Renderer[FieldDefinition] {
+    override def unsafeRender(definition: FieldDefinition, indent: Option[Int], builder: StringBuilder): Unit =
+      definition match {
+        case FieldDefinition(description, name, arguments, tpe, directives) =>
+          builder += '\n'
+          unsafeRenderDescription(description, indent, builder)
+          pad(indent, builder)
+          builder ++= name
+          unsafeRenderArguments(arguments, builder)
+          builder ++= s": $tpe"
+
+          directivesRenderer.unsafeRender(directives, None, builder)
+      }
+  }
+
+  private def unsafeRenderArguments(definitions: List[InputValueDefinition], builder: StringBuilder): Unit =
+    if (definitions.nonEmpty) {
+      builder += '('
+      var first = true
+      definitions.foreach { definition =>
+        if (first) first = false
+        else builder += ','
+        unsafeRenderDescription(definition.description, None, builder, newline = false)
+        builder ++= definition.name
+        builder ++= ": "
+        builder ++= innerType(definition.ofType)
+        definition.defaultValue.foreach { value =>
+          builder ++= " = "
+          builder ++= value.toInputString
+        }
+        directivesRenderer.unsafeRender(definition.directives, None, builder)
+      }
+      builder += ')'
+    }
+
+  private def pad(indentation: Option[Int], writer: StringBuilder): Unit = {
+    var i = indentation.getOrElse(0)
+    while (i > 0) {
+      writer ++= "  "
+      i -= 1
+    }
+  }
+
+  private def increment(indentation: Option[Int]): Option[Int] = indentation.map(_ + 1)
+
+  private def unsafeFastString(value: String, builder: StringBuilder) = {
+    builder.append('"')
+    var i = 0
+    while (i < value.length) {
+      value.charAt(i) match {
```

Comment:
I believe that in most cases the char will be the default `c`, is that right? If that's the case, consider using a lookup switch (binary search) so that we check only 3 (maybe 4) of the conditions instead of all 7

```suggestion
      (value.charAt(i): @annotation.switch) match {
```

Reply:
Ah nice!